File: backend/app/schemas/report_card_schema.py
# backend/app/schemas/report_card_schema.py

import uuid
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict


class SubjectMark(BaseModel):
    """
    Represents the marks for a single subject within one exam.
    """

    subject_name: str
    marks_obtained: Decimal
    max_marks: Decimal

    model_config = ConfigDict(from_attributes=True)


class ExamSummary(BaseModel):
    """
    Represents the summary for a single exam, including all its subjects
    and the calculated totals and percentage for that exam.
    """

    exam_name: str
    marks: list[SubjectMark]
    total_obtained: Decimal
    total_max_marks: Decimal
    percentage: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class ReportCard(BaseModel):
    """
    The main report card model. It contains student info, a list
    of all their exam summaries, and the grand totals.
    """

    student_id: int
    student_user_id: uuid.UUID
    student_name: str
    class_name: str
    exam_summaries: list[ExamSummary]
    grand_total_obtained: Decimal
    grand_total_max_marks: Decimal
    overall_percentage: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)

File: apps/admin-web/src/app/services/reportcard.schema.ts
import { z } from "zod";

/**
 * Subject performance in a report card
 */
export const SubjectPerformance = z.object({
  subject_id: z.number(),
  subject_name: z.string(),
  marks_obtained: z.number().nullable(),
  marks_total: z.number(),
  percentage: z.number().nullable(),
  grade: z.string().nullable(),
  remarks: z.string().nullable().optional(),
});
export type SubjectPerformance = z.infer<typeof SubjectPerformance>;

/**
 * Student report card
 */
export const ReportCard = z.object({
  student_id: z.number(),
  student_name: z.string(),
  class_name: z.string().optional(),
  academic_year_id: z.number(),
  academic_year_name: z.string().optional(),
  subjects: z.array(SubjectPerformance),
  total_marks_obtained: z.number(),
  total_marks_possible: z.number(),
  overall_percentage: z.number(),
  overall_grade: z.string().nullable(),
  attendance_percentage: z.number().nullable().optional(),
  teacher_remarks: z.string().nullable().optional(),
  generated_at: z.string().optional(),
});
export type ReportCard = z.infer<typeof ReportCard>;
